core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/SamplesReceivedWorkflow.java

```diff
@@ -0,0 +1,536 @@
+        .collect(Collectors.toList());
+  }
+
+  private boolean hasInput(WorkflowStep step) {
+    return step.getProgressStep() != null;
+  }
+
+  @Override
+  protected WorkflowName getWorkflowName() {
+    return WorkflowName.SAMPLES_RECEIVED;
+  }
+
+  @Override
+  public WorkflowStepPrompt getStep(int stepNumber) {
+    if (stepNumber == 0) {
+      return sampleStep.getPrompt();
+    } else if (stepNumber == 1) {
+      return qcStep.getPrompt();
+    } else if (stepNumber == 2) {
+      return aliquotStep.getPrompt();
+    } else if (stepNumber == 3) {
+      return boxStep.getPrompt();
+    } else {
+      return aliquotHandlingSteps.get(asAliquotIndex(stepNumber)).getPrompt();
+    }
+  }
+
+  private int asAliquotIndex(int stepNumber) {
+    return stepNumber - 4;
+  }
+
+  @Override
+  public boolean isComplete() {
+    return Stream.concat(Stream.of(sampleStep, qcStep, aliquotStep, boxStep), aliquotHandlingSteps.stream()).allMatch(this::hasInput);
+  }
+
+  @Override
+  public List<String> processInput(int stepNumber, ProgressStep step) {
+    List<String> errors = new ArrayList<>();
+    if (stepNumber == 0) {
+      step.accept(sampleStep);
+      qcStep.cancelInput();
+      aliquotStep.cancelInput();
+      boxStep.cancelInput();
+      cancelAliquotHandlingSteps();
+    } else if (stepNumber == 1) {
+      step.accept(qcStep);
+      aliquotStep.cancelInput();
+      boxStep.cancelInput();
+      cancelAliquotHandlingSteps();
+    } else if (stepNumber == 2) {
+      step.accept(aliquotStep);
+      boxStep.cancelInput();
+      boxStep.setNumAliquots(aliquotStep.getAliquotQuantity());
+      cancelAliquotHandlingSteps();
+    } else if (stepNumber == 3) {
+      step.accept(boxStep);
+      if (hasInput(boxStep)) {
+        resetAliquotHandlingSteps(aliquotStep.getAliquotQuantity(), boxStep.getBox());
+      } else {
+        errors.add(boxStep.getError());
+      }
+    } else {
+      step.accept(aliquotHandlingSteps.get(asAliquotIndex(stepNumber)));
+      if (aliquotHandlingSteps.get(asAliquotIndex(stepNumber)).getError() != null) {
+        errors.add(aliquotHandlingSteps.get(asAliquotIndex(stepNumber)).getError());
+      }
+    }
+    return errors;
+  }
+
+  private void cancelAliquotHandlingSteps() {
+    for (WorkflowStep aliquotHandlingStep : aliquotHandlingSteps) {
+      aliquotHandlingStep.cancelInput();
+    }
+  }
+
+  private void resetAliquotHandlingSteps(int aliquotQuantity, Box box) {
+    aliquotHandlingSteps = new ArrayList<>();
+    for (int i = 0; i < aliquotQuantity; i++) {
+      aliquotHandlingSteps.add(new BoxPositionStep(i, box, occupiedLocations));
+      aliquotHandlingSteps.add(new VolumeStep(i));
+      aliquotHandlingSteps.add(new ConcentrationStep(i));
+    }
+  }
+
+  @Override
+  public void cancelInput() {
+    if (aliquotHandlingSteps.stream().anyMatch(this::hasInput)) {
+      getLastCompletedAliquotHandlingStep().cancelInput();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "getLastCompletedAliquotHandlingStep()" can return null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)